### ISSUE TITLE
FindProperty(<Name>).Relational().ColumnType is null if column type is not specifically set

### ISSUE THREAD

**Issue by JonPSmith** [aspnet/EntityFramework#8034]
Describe what is not working as expected.

I am trying to obtain the database column type by using `entityType.FindProperty(<PropertyName>).Relational()` and I get null for `ColumnType` for string properties. If I use the same command on another, ~~non-string property it works.~~ property where I have set the type then it contains that type.

### Steps to reproduce
Include a complete code listing (or project/solution) that we can run to reproduce the issue.

Unit Test
```c#
[Fact]
public void TestGetColumnRelationalTypeOk()
{
    //SETUP
    var inMemDb = new SqliteInMemory();

    //ATTEMPT
    using (var context = inMemDb.GetContextWithSetup())
    {
        var colType = context.GetColumnRelationalType(new Book(), p => p.Description);

        //VERIFY
        colType .ShouldEqual("nvarchar(max)");
    }
}
```
The `GetColumnRelationalType` extension method is
```c#
public static string GetColumnRelationalType<TEntity, TProperty>(this DbContext context,
    TEntity source, Expression<Func<TEntity, TProperty>> model) where TEntity : class
{
    var efType = context.Model.FindEntityType(typeof(TEntity).FullName);
    var propInfo = GetPropertyInfoFromLambda(model);
    var relational = efType.FindProperty(propInfo.Name).Relational();
    return relational.ColumnType;
}
```
My Book Entity is 
```c#
public class Book                        
{
    public const int WebUrlMaxlength = 512; 

    public int BookId { get; set; }

    [Required] 
    [MaxLength(256)] 
    public string Title { get; set; }
    public string Description { get; set; }
    public DateTime PublishedOn { get; set; }
    [MaxLength(64)] 
    public string Publisher { get; set; }
    public decimal Price { get; set; }

    [MaxLength(WebUrlMaxlength)] 
    public string ImageUrl { get; set; }

    //----------------------------------------------
    //relationships

    public PriceOffer Promotion { get; set; }       
    public ICollection<Review> Reviews { get; set; }
    public ICollection<BookAuthor> 
        AuthorsLink { get; set; }                   
}
```
I haven't included my DbContext, but it is a standard layout with a `DbSet<Book> Books` property.

I have tried this with in-memory Sqlite and with SQL Server - both return null.

### Further technical details
EF Core version: 1.1.1
Database Provider:Tried with both Microsoft.EntityFrameworkCore.Sqlite and Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10
IDE:  Visual Studio 2017

**Comment by ajcvickers** [comment]
@JonPSmith This looks like a dupe of #7351, for which we are also planning to add some sugar API in #7461.

However, I am surprised that "If I use the same command on another, non-string property it works." I have not been able to reproduce this. The only time ColumnType should be non-null is when the store type has been explicitly set in the model.

**Comment by JonPSmith** [comment]
Hi @ajcvickers,

Thanks for that. I was obviously using the wrong approach to getting the store type, and I have changed to the procedure in #7351. One further question: how can I convert the type that `typeMapper.FindMapping(efProperty)` returns into the actual data type that the database provider will use? 

**Feedback on the API:** 
I support the discussion in #7461, and here are my comments.

`FindProperty(<Name>).Relational().ColumnName` provides the column name, even if default, as does `IEntityType.Relational().TableName`, while `FindProperty(<Name>).Relational().ColumnType` only returns the type if it has been specifically set. For me the `.ColumnType` is out of step with the rest of the interfaces. 

Also, if `FindProperty(<Name>).Relational().ColumnType` returned the generic column type, e.g. TEXT for strings, then  `FindProperty(<Name>).Sqlite().ColumnType` etc. could returns the actual column type, e.g. varchar or nvarchar, for the specific database provider. That would make decoding specific databases much easier.

**NOTE:** 

>  I am surprised that "If I use the same command on another, non-string property it works."

I just happened to pick a property where I had set the type directly. Sorry, I should have included the application's DbContext.

**Comment by ajcvickers** [comment]
@JonPSmith There are a couple of things going on here:
* Relational() will only return values set generally for all relational providers, and will not set values set explicitly for only one provider. This is what you were referring to with the ForSqlServer... comments above. For this, reason, Relational() should not be used for consuming metadata. We did in fact remove those methods to stop people doing this (see #2701 by me with @AndriySvyryd signing off) but unfortunately they got added back in before we shipped (see #5482 ironically by @AndriySvyryd with me signing off). However, this all changes in 2.0 anyway because of #7166.
* Some store metadata is defined entirely by the model. This is the case for ColumnName, for example. Other store metadata is computed by the runtime and can change based on the context in which the model is being used. Column types are like that. In this case the metadata stored on the model is only there when set explicitly in the app. I agree that to the consumer these look the same, which can be confusing. We will discuss this for 2.0.

It's probably also worth reading this thread on consuming metadata: #7725

**Comment by ajcvickers** [comment]
We will look at making the API a bit easier to understand, possibly by changing the names of some properties in 2.0 to more accurately reflect that they only contain explicitly set values and will otherwise be nulll.

**Comment by JonPSmith** [comment]
I agree - that is the most sensible approach.

Here are my **use cases** for finding out information on the database. I have tried to put them in some sort of order, with what I think are thee more common usages earlier.

1. Find out a table name. I have used this in code to do a fast clear of a table use raw SQL via ADO.NET and I have seen other do that too. Also useful if building raw ADO.NET SQL commands, as the table names are variable: property name or class name. (Column names aren't so hard, as they normally follow the property name).
2. Find out what properties are keys. (I use this in [GenericServices](https://github.com/JonPSmith/GenericServices) to copy the keys to/from the DTO. I expect other do this too).
3. Find out everything about the the database to do a compare between EF Model and actual database and show the differences. (I use this in [EfSchemaCompare](https://github.com/JonPSmith/EfSchemaCompare). I don't expect many people will do that).

I hope that helps.
